### ISSUE TITLE
Possible bug in Legion Prof+Spy

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1330]
I'm running this program with detailed Legion Spy enabled:

https://gitlab.com/StanfordLegion/legion/-/blob/legion_prof_rs_spy/language/tests/regent/run_pass/spy10.rg

Here's a zoomed in view of the profiles:

  * Python: http://sapling.stanford.edu/~eslaught/legion_prof_spy_comparison/legion_prof_spy_py/?start=10065084.96862559&end=10079843.008492218&collapseAll=false&resolution=10
  * Rust: http://sapling.stanford.edu/~eslaught/legion_prof_spy_comparison/legion_prof_spy_rs/?start=10065084.96862559&end=10079843.008492218&collapseAll=false&resolution=10

If you click on the task `asdf` in Python, you see it has no task dependencies. In Rust, it has dependencies on the two `hello` tasks.

Strictly speaking, I think Rust is more correct. The `hello` tasks literally do have dependencies on everything that came before them due to the execution fence.

An even simpler program with a fence shows the dependency with both Python and Rust:

https://gitlab.com/StanfordLegion/legion/-/blob/legion_prof_rs_spy/language/tests/regent/run_pass/spy9.rg

  * Python: http://sapling.stanford.edu/~eslaught/legion_prof_spy_comparison_spy9/legion_prof_spy_py/?start=10065251.716295231&end=10069488.246040273&collapseAll=false&resolution=10
  * Rust: http://sapling.stanford.edu/~eslaught/legion_prof_spy_comparison_spy9/legion_prof_spy_rs/?start=10065251.716295231&end=10069488.246040273&collapseAll=false&resolution=10

That seems to support my case that fence dependencies should be tracked.

All tests were done in `legion_prof_rs_spy` (based on a slightly dated version of `master`).

**Comment by lightsighter** [comment]
Do you think this is a bug in the Legion Prof side of the code or the Legion Spy side of the code? I would be quite surprised to see Legion Spy missing a calculation of a fence dependence.

**Comment by elliottslaughter** [comment]
I really don't have a sense for this. Is there a way to figure out which is responsible?

I have observed that `legion_prof.py` does take the Spy output and post-process it, so it may be it's doing something wrong here:

https://github.com/StanfordLegion/legion/blob/2370b143b701fc714a99da6a2c7ab98dc0784d0a/tools/legion_prof.py#L4422-L4434

(And this continues for the next ~50 lines or so.)

**Comment by lightsighter** [comment]
These aren't even the right dependences to be capturing for looking for a critical path. Legion Prof should be recording the Realm event graph, but these are just relationships between point tasks, slice tasks, and index tasks. I'm not sure how Legion Prof is getting any kind of dependence right here. The comments suggest the code is also looking for the operation dependences (but those would be the mapping dependences, not the physical event dependences), and would also be wrong.

> Is there a way to figure out which is responsible?

Verification of the detailed legion spy logging would fail if it is Legion Spy's fault that the dependences are missing (either because the runtime didn't compute them, didn't log them, or Legion Spy didn't parse them).

**Comment by elliottslaughter** [comment]
I am definitely not an expert in this code, but I believe the dependencies come from the code below. What I linked you earlier was (I think) used only for parent-child relationships.

https://github.com/StanfordLegion/legion/blob/2370b143b701fc714a99da6a2c7ab98dc0784d0a/tools/legion_prof.py#L4499-L4520

That seems to be using a Legion Spy API called `print_event_graph`, which certainly sounds promising. But I really don't understand the subtleties of these things.

**Comment by lightsighter** [comment]
> What I linked you earlier was (I think) used only for parent-child relationships.

No, that code was for matching point tasks up with the index task owners through the slice tasks.

> I am definitely not an expert in this code, but I believe the dependencies come from the code below.

This code looks better. It's using the `print_event_graph` function to gather the names of all the nodes in the event graph, and then using the `physical_incoming` member on each of those nodes to get the edges to the previous nodes. I would be pretty surprised if this didn't include the transitive edges for the fences. Is the problem that the edges are missing entirely, or they just don't look the way you want? (I actually can't see anything wrong in the profiles that I looked at above.) Also was the event graph computed by Legion Spy or was it just recorded from the runtime?

**Comment by elliottslaughter** [comment]
The dependency is missing outright, as described in the original issue:

> If you click on the task `asdf` in Python, you see it has no task dependencies. In Rust, it has dependencies on the two `hello` tasks.

This is detailed logging. You can find the log file at: http://sapling.stanford.edu/~eslaught/legion_prof_spy_comparison/spy_0.log

**Comment by elliottslaughter** [comment]
If you go back to the original logs, you'll need to check out Legion at b9d2d9ba66458fad633f9b7bbfd40b7d9e8fa4a7 to be able to parse them.

I think this is a `legion_prof.py` bug. `legion_spy.py -e` generates what you'd expect:

[event_graph_main_1.pdf](https://github.com/StanfordLegion/legion/files/9642930/event_graph_main_1.pdf)

(And I confirmed by rerunning `legion_prof.py` on those logs that it still generates the same output.)

**Comment by lightsighter** [comment]
Right, that is the event graph that I would expect Legion Spy to generate, so I think it's behavior is correct. Most likely Legion Prof is not picking up on some aspect of the dependence analysis graph (unclear why).
